library/psa_crypto.c

```diff
@@ -4111,14 +4125,123 @@ static psa_status_t psa_key_derivation_tls12_prf_generate_next_block(
+     *
+     * A(0) = seed
+     * A(i) = HMAC_hash(secret, A(i-1))
+     *
+     * The `psa_tls12_prf_key_derivation` structure saves the block
+     * `HMAC_hash(secret, A(i) + seed)` from which the output
+     * is currently extracted as `output_block`.
+     */
+
+    /* Save the hash context before using it, to preserve the hash state with
+     * only the inner padding in it. We need this, because inner padding depends
+     * on the key (secret in the RFC's terminology). */
+    status = psa_hash_clone( &tls12_prf->hmac.hash_ctx, &backup );
+    if( status != PSA_SUCCESS )
+        goto cleanup;
+
+    /* Calculate A(i) where i = tls12_prf->block_number. */
+    if( tls12_prf->block_number == 1 )
+    {
+        /* A(1) = HMAC_hash(secret, A(0)), where A(0) = seed. (The RFC overloads
+         * the variable seed and in this instance means it in the context of the
+         * P_hash function, where seed = label + seed.) */
+        status = psa_hash_update( &tls12_prf->hmac.hash_ctx,
+                                  tls12_prf->label, tls12_prf->label_length );
+        if( status != PSA_SUCCESS )
+            goto cleanup;
+        status = psa_hash_update( &tls12_prf->hmac.hash_ctx,
+                                  tls12_prf->seed, tls12_prf->seed_length );
+        if( status != PSA_SUCCESS )
+            goto cleanup;
+    }
+    else
+    {
+        /* A(i) = HMAC_hash(secret, A(i-1)) */
+        status = psa_hash_update( &tls12_prf->hmac.hash_ctx,
+                                  tls12_prf->Ai, hash_length );
+        if( status != PSA_SUCCESS )
+            goto cleanup;
+    }
+
+    status = psa_hmac_finish_internal( &tls12_prf->hmac,
+                                       tls12_prf->Ai, hash_length );
+    if( status != PSA_SUCCESS )
+        goto cleanup;
+    status = psa_hash_clone( &backup, &tls12_prf->hmac.hash_ctx );
+    if( status != PSA_SUCCESS )
+        goto cleanup;
+
+    /* Calculate HMAC_hash(secret, A(i) + label + seed). */
+    status = psa_hash_update( &tls12_prf->hmac.hash_ctx,
+                              tls12_prf->Ai, hash_length );
+    if( status != PSA_SUCCESS )
+        goto cleanup;
+    status = psa_hash_update( &tls12_prf->hmac.hash_ctx,
+                              tls12_prf->label, tls12_prf->label_length );
+    if( status != PSA_SUCCESS )
+        goto cleanup;
+    status = psa_hash_update( &tls12_prf->hmac.hash_ctx,
+                              tls12_prf->seed, tls12_prf->seed_length );
+    if( status != PSA_SUCCESS )
+        goto cleanup;
+    status = psa_hmac_finish_internal( &tls12_prf->hmac,
+                                       tls12_prf->output_block, hash_length );
+    if( status != PSA_SUCCESS )
+        goto cleanup;
+    status = psa_hash_clone( &backup, &tls12_prf->hmac.hash_ctx );
+    if( status != PSA_SUCCESS )
+        goto cleanup;
+
+
+cleanup:
+
+    cleanup_status = psa_hash_abort( &backup );
```

Comment:
`cleanup_status` is only used here. How about:
```suggestion
     if( status == PSA_SUCCESS )
         status = psa_hash_abort( &backup );
     else
         psa_hash_abort( &backup );
         
         return( status );
         
     
        status = cleanup_status;
```

Reply:
I thought about that, but my intuition was that it would take more code size. Now I wrote a toy example to compare them and this is what I've got:

With temporary variable:
```asm
00000000 <test1>:
   0:   b510            push    {r4, lr}
   2:   0004            movs    r4, r0
   4:   4803            ldr     r0, [pc, #12]   ; (14 <test1+0x14>)
   6:   f7ff fffe       bl      0 <printf>
   a:   2c00            cmp     r4, #0
   c:   d000            beq.n   10 <test1+0x10>
   e:   0020            movs    r0, r4
  10:   bd10            pop     {r4, pc}
  12:   46c0            nop                     ; (mov r8, r8)
  14:   00000000        .word   0x00000000
```

And without:
```asm
00000018 <test2>:
  18:   b510            push    {r4, lr}
  1a:   1e04            subs    r4, r0, #0
  1c:   d105            bne.n   2a <test2+0x12>
  1e:   4805            ldr     r0, [pc, #20]   ; (34 <test2+0x1c>)
  20:   f7ff fffe       bl      0 <printf>
  24:   0004            movs    r4, r0
  26:   0020            movs    r0, r4
  28:   bd10            pop     {r4, pc}
  2a:   4803            ldr     r0, [pc, #12]   ; (38 <test2+0x20>)
  2c:   f7ff fffe       bl      0 <puts>
  30:   e7f9            b.n     26 <test2+0xe>
  32:   46c0            nop                     ; (mov r8, r8)
  34:   00000000        .word   0x00000000
  38:   00000006        .word   0x00000006
```

The temporary variable seems to save 3 instructions (6 byte of code size when using thumb, 12 bytes otherwise). Therefore I think it is better to keep this as it is.